File: 2019/day18/cpp/src/maze.cpp
#include "maze.hpp"
#include "utils.hpp"
#include <queue>

// Data structure for the queue
struct queue_data
{
  int m_depth;
  std::vector<std::pair<int, int>> m_pos;
  std::set<char> m_keys;

  queue_data(int n1, std::vector<std::pair<int, int>> n2, std::set<char> n3) : 
      m_depth(n1), m_pos(n2), m_keys(n3) {}

  bool operator<(const struct queue_data &other) const {
    return m_depth > other.m_depth;
  }
};

Maze::Maze(std::string file_name, bool multiple_robots) : 
    m_grid(), m_y(), m_x(), m_total_keys(), m_pos() {

  utils::readFile(file_name,m_grid);
  find_start_and_keys();
  m_pos = initialize_position(multiple_robots);

}

// Go through the maze and find the starting position + all keys
void Maze::find_start_and_keys() {

  auto width = m_grid[0].size();

  std::string linear_grid = "";
  for (auto &str : m_grid) {
    linear_grid += str;
  }

  // Start position
  auto start_index = linear_grid.find('@');
  m_y = start_index / width;
  m_x = start_index % width;

  // Store all lower case keys
  for (auto &chr : linear_grid) {
    if (islower(chr)) {
      m_total_keys.insert(chr);
    }
  }

}

std::vector<std::pair<int, int>> Maze::initialize_position(bool multi_robots) {

  // Reshape the map if there are multiple robots
  auto len = m_grid[0].size();
  if (multi_robots) {
    m_grid[m_y-1] = m_grid[m_y-1].substr(0,m_x) + "#" + 
                    m_grid[m_y-1].substr(m_x+1, len-(m_x+1));
    m_grid[m_y]   = m_grid[m_y].substr(0, m_x-1) + "###" + 
                    m_grid[m_y].substr(m_x + 2, len - (m_x + 2));
    m_grid[m_y+1] = m_grid[m_y+1].substr(0, m_x) + "#" + 
                    m_grid[m_y+1].substr(m_x + 1, len - (m_x + 1));

    std::pair<int,int> p1(m_y-1, m_x-1), p2(m_y+1, m_x+1);
    std::pair<int,int> p3(m_y+1, m_x-1), p4(m_y-1, m_x+1);

    return std::vector<std::pair<int,int>>({p1,p2,p3,p4});

  // Otherwise return positions
  } else {
    return std::vector<std::pair<int,int>>({std::make_pair(m_y,m_x)});
  }
}

std::vector<std::tuple<int, int, int, char>> 
    Maze::reachable_keys(int y, int x, std::set<char> keys) {

  // Depth, y, x, key
  std::vector<std::tuple<int,int,int,char>> dist_to_key;
  std::queue<std::tuple<int,int,int>> q;
  q.push(std::make_tuple(y,x,0));
  
  // Keeping track of visited nodes
  std::set<std::pair<int,int>> visited;

  // Search directions
  std::vector<std::vector<int>> directions({ {1,0},{-1,0},{0,1},{0,-1}});

  // Keep doing until we can't go no more 
  while (!q.empty()) {
    auto q_ = q.front(); q.pop();
    auto y_ = std::get<0>(q_);
    auto x_ = std::get<1>(q_);
    auto depth = std::get<2>(q_);

    // If the grid elemnt is a key and we dont have it
    if (islower(m_grid[y_][x_]) and !keys.count(m_grid[y_][x_])) {
      dist_to_key.push_back(std::make_tuple(depth, y_, x_, m_grid[y_][x_]));
      continue; // Maybe not here
    }

    for (auto &dir : directions) {
      auto y_new = dir[0] + y_;
      auto x_new = dir[1] + x_;

      // Skip if we've been here
      if (visited.count(std::make_pair(y_new,x_new))) {
        continue;
      }
      visited.insert(std::make_pair(y_new,x_new));

      // Keep going in new direction
      auto value = m_grid[y_new][x_new];
      if (value != '#' && (!isupper(value) || keys.count(tolower(value)))) {
        q.push(std::make_tuple(y_new,x_new,depth+1));
      }
    }
  }
  return dist_to_key;
}

int Maze::compute_shortest_path() {

  // Go out and collect keys
  std::priority_queue<queue_data> q_;
  q_.push(queue_data(0,m_pos,std::set<char>()));

  // Visited nodes
  std::set< std::pair<std::vector<std::pair<int,int>>,std::set<char>>> visited;

  while (!q_.empty()) {

    // Get node with smallest depth
    queue_data q = q_.top(); q_.pop();

    // If we have all keys we are done
    auto keys = q.m_keys;
    if (keys == m_total_keys) {
      return q.m_depth;
    }

    // Skip if we've been here
    if (visited.count(std::make_pair(q.m_pos,q.m_keys))) {
      continue;
    }
    visited.insert(std::make_pair(q.m_pos,q.m_keys));

    // Loop through the positions (if multiple robots)
    for (size_t i = 0; i < q.m_pos.size(); i++) {
      // Get all reachable keys
      auto rk = reachable_keys(q.m_pos[i].first, q.m_pos[i].second, q.m_keys);

      // Go through all keys we can reach
      for (size_t j = 0; j < rk.size(); j++) {
        int d = std::get<0>(rk[j]);
        int y = std::get<1>(rk[j]), x = std::get<2>(rk[j]);
        char keyy = std::get<3>(rk[j]);
        
        // Update the new position
        auto new_pos = q.m_pos;
        new_pos[i] = std::make_pair(y,x);
        auto setkey = q.m_keys; setkey.insert(keyy);
        q_.push(queue_data(q.m_depth+d,new_pos,setkey));
      }
    } 
  }
  return -1;
}

File: 2019/day18/python/maze.py
import collections
import itertools
import heapq


class Maze:
    """Class finding its way through the maze."""

    def __init__(self, fileName, multi_robots=False):
        self.grid = self.read_map(fileName)
        self.y, self.x, self.total_keys = self.find_start_and_keys()
        self.pos = self.initialize_position(multi_robots)

    def compute_shortest_path(self):
        """Traverse the grid by first looking for the reachable keys and select
        nodes with shortest distance while traversing."""

        # Initial node - depth, position, collection of keys
        q = [(0, self.pos, frozenset())]

        # Visited nodes
        visited = set()

        while q:

            # Obtain shortest the node
            depth, current_pos, keys = heapq.heappop(q)

            # We are done if all keys are found
            if keys == self.total_keys:
                return depth

            # Skip node if visited before
            if (current_pos, keys) in visited:
                continue

            # Update visited node
            visited.add((current_pos, keys))

            # Go through positions
            for i, (current_y, current_x) in enumerate(current_pos):
                # Go through the reachable keys for each position
                for d, y, x, key in self.reachable_keys(
                    current_y, current_x, keys
                ):
                    # Update positions list
                    new_pos = (
                        current_pos[0:i] + ((y, x),) + current_pos[i + 1 :]
                    )
                    # Push to heap new depth, new position, and collected keys
                    heapq.heappush(
                        q, (depth + d, new_pos, keys | frozenset([key]))
                    )

    def reachable_keys(self, y_, x_, keys):
        """Given start position y_,x_, and keys find other reachable keys"""

        # Initialize node to go through
        q = collections.deque([(y_, x_, 0)])

        # Initialize visited nodes
        visited = set()

        # Search directions
        directions = ((1, 0), (-1, 0), (0, 1), (0, -1))

        while q:

            # Get the node
            y, x, depth = q.popleft()

            # If we reach a key which we do not have, yield it
            if self.grid[y][x].islower() and self.grid[y][x] not in keys:
                yield depth, y, x, self.grid[y][x]
                continue

            # Search neighbors
            for dy, dx in directions:
                y_new, x_new = y + dy, x + dx

                # Skip if we have visited the node
                if ((y_new, x_new)) in visited:
                    continue

                # Add to visited node
                visited.add((y_new, x_new))

                value = self.grid[y_new][x_new]

                if value != "#" and (
                    not value.isupper() or value.lower() in keys
                ):
                    q.append((y_new, x_new, depth + 1))

    def initialize_position(self, multi_robots):
        """Set the initial position depending on part one or part two"""
        if multi_robots:
            y, x = self.y, self.x
            self.grid[y - 1] = (
                self.grid[y - 1][:x] + "#" + self.grid[y - 1][x + 1 :]
            )
            self.grid[y] = (
                self.grid[y][: x - 1] + "###" + self.grid[y][x + 2 :]
            )
            self.grid[y + 1] = (
                self.grid[y + 1][:x] + "#" + self.grid[y + 1][x + 1 :]
            )
            return (
                (self.y - 1, self.x - 1),
                (self.y - 1, self.x + 1),
                (self.y + 1, self.x - 1),
                (self.y + 1, self.x + 1),
            )
        else:
            return ((self.y, self.x),)

    def find_start_and_keys(self):
        """Figure out middle of the map and how many keys are present"""
        # Flatten grid
        linear_grid = list(itertools.chain.from_iterable(self.grid))

        # Width and height of grid
        width = len(self.grid[0])
        start_index = linear_grid.index("@")

        # Divide index with length and width
        y, x = start_index // width, start_index % width

        # If the key is a lower case letter then include
        total_keys = set(key for key in linear_grid if key.islower())

        return y, x, total_keys

    def read_map(self, fileName):
        """Read the map"""
        with open(fileName) as f:
            return f.read().splitlines()

    def print_map(self):
        """Print the map"""
        for row in self.grid:
            print("".join(row))
